icechunk/src/store.rs

```diff
@@ -337,47 +337,60 @@ impl Store {
+    pub async fn delete_dir(&self, prefix: &str) -> StoreResult<()> {
         if self.read_only().await {
             return Err(StoreError::ReadOnly);
         }
 
-        match Key::parse(key)? {
-            Key::Metadata { node_path } => {
-                // we need to hold the lock while we do the node search and the write
-                // to avoid race conditions with other writers
-                // (remember this method takes &self and not &mut self)
-                let mut guard = self.session.write().await;
-                let node = guard.get_node(&node_path).await;
-
-                // When there is no node at the given key, we don't consider it an error, instead we just do nothing
-                if let Err(SessionError::NodeNotFound { path: _, message: _ }) = node {
-                    return Ok(());
-                };
-
-                let node = node.map_err(StoreError::SessionError)?;
-                match node.node_data {
-                    NodeData::Array(_, _) => {
-                        Ok(guard.deref_mut().delete_array(node_path).await?)
-                    }
-                    NodeData::Group => {
-                        Ok(guard.deref_mut().delete_group(node_path).await?)
-                    }
-                }
-            }
-            Key::Chunk { node_path, coords } => {
-                let mut guard = self.session.write().await;
-                let session = guard.deref_mut();
-                match session.set_chunk_ref(node_path, coords, None).await {
-                    Ok(_) => Ok(()),
-                    Err(SessionError::NodeNotFound { path: _, message: _ }) => {
-                        // When there is no chunk at the given key, we don't consider it an error, instead we just do nothing
-                        Ok(())
+        // TODO: Handling preceding "/" is ugly!
+        let path = format!("/{}", prefix.trim_start_matches("/"))
+            .try_into()
+            .map_err(|_| StoreError::BadKeyPrefix(prefix.to_owned()))?;
+
+        {
+            let mut guard = self.session.write().await;
+            let node = guard.get_node(&path).await;
+            match node {
+                Ok(NodeSnapshot {
+                    node_data: NodeData::Group, path: node_path, ..
+                }) => Ok(guard.deref_mut().delete_group(node_path).await?),
+                Ok(NodeSnapshot {
+                    node_data: NodeData::Array(..),
+                    path: node_path,
+                    ..
+                }) => Ok(guard.deref_mut().delete_array(node_path).await?),
+                Err(SessionError::NodeNotFound { .. }) => {
+                    // other cases are
+                    // 1. delete("/path/to/array/c")
+                    // 2. delete("/path/to/array/c/0/0")
+                    let prefix = prefix.trim_start_matches("/");
+                    let to_delete = guard
+                        .deref_mut()
+                        .all_chunks()
```

Comment:
Some thoughts:
1. This is only deleting chunks, isn't it? How about nodes?
2. If we are going to delete only chunks, we should be able to do it without all the extra `get_node` calls involved in `delete_key`
3. We should find a way to do a big delete without gathering every chunk into a vector in memory at once
4. It feels like we should have a way to get nodes based on a prefix, and then use the high level operations that delete nodes